internal/batch/batch_processor.go

```diff
@@ -101,103 +130,245 @@ func (bw *batchWork) estimateSize() int64 {
+	newQueue := make([]*batchWork, 0, len(bp.assemblyQueue)+1)
+	added := false
+	// Check it's not in the recently flushed list
+	for _, flushedSequence := range bp.flushedSequences {
+		if newWork.msg.Sequence == flushedSequence {
+			log.L(bp.ctx).Debugf("Ignoring add of recently flushed message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			return false, false
+		}
+	}
+	// Build the new sorted work list, checking there for duplicates too
+	for _, work := range bp.assemblyQueue {
+		if newWork.msg.Sequence == work.msg.Sequence {
+			log.L(bp.ctx).Debugf("Ignoring duplicate add of message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+			return false, false
+		}
+		if !added && newWork.msg.Sequence < work.msg.Sequence {
+			newQueue = append(newQueue, newWork)
+			added = true
+		}
+		newQueue = append(newQueue, work)
+	}
+	if !added {
+		newQueue = append(newQueue, newWork)
+	}
+	log.L(bp.ctx).Debugf("Added message %s sequence=%d to in-flight batch assembly %s", newWork.msg.Header.ID, newWork.msg.Sequence, bp.assemblyID)
+	bp.assemblyQueueBytes += newWork.estimateSize()
+	bp.assemblyQueue = newQueue
+	full = len(bp.assemblyQueue) >= int(bp.conf.BatchMaxSize) || (bp.assemblyQueueBytes >= bp.conf.BatchMaxBytes)
+	overflow = len(bp.assemblyQueue) > 1 && (bp.assemblyQueueBytes > bp.conf.BatchMaxBytes)
+	return full, overflow
+}
+
+func (bp *batchProcessor) startFlush(overflow bool) (id *fftypes.UUID, flushAssembly []*batchWork, byteSize int64) {
+	bp.statusMux.Lock()
+	defer bp.statusMux.Unlock()
+	// Start the clock
+	bp.flushStatus.Blocked = false
+	bp.flushStatus.LastFlushTime = fftypes.Now()
+	// Split the current work if required for overflow
+	overflowWork := make([]*batchWork, 0)
+	if overflow {
+		lastElem := len(bp.assemblyQueue) - 1
+		flushAssembly = append(flushAssembly, bp.assemblyQueue[:lastElem]...)
+		overflowWork = append(overflowWork, bp.assemblyQueue[lastElem])
+	} else {
+		flushAssembly = bp.assemblyQueue
+	}
+	// We need to keep track of the sequences we're flushing, because until we finish our flush
+	// the batch processor might be re-queuing the same messages to use due to rewinds.
+	// We keep all of the last batch, and up to twice the batch size over time (noting our channel
+	// size is our batch size - so the batch manager cannot get further than that ahead).
+	newFlushedSeqLen := len(flushAssembly) + len(bp.flushedSequences)
+	maxFlushedSeqLen := int(2 * bp.conf.BatchMaxSize)
+	if newFlushedSeqLen > maxFlushedSeqLen && maxFlushedSeqLen > len(bp.flushedSequences) {
+		newFlushedSeqLen = maxFlushedSeqLen
+	}
+	newFlushedSequences := make([]int64, newFlushedSeqLen)
+	for i := 0; i < len(flushAssembly); i++ {
+		// Add in reverse order - so we can trim the end of it off later (in the next block) and keep the newest
+		newFlushedSequences[len(flushAssembly)-i-1] = flushAssembly[i].msg.Sequence
```

Comment:
So... we are adding to the front of a list, in reverse order, and then ultimately trimming the end of the list.

It feels like we could add to the end of the list in order, and then trim the front, and it might be more understandable. However, I do think the current logic works, so it's not mandatory to re-spin this again if we want to defer for now.